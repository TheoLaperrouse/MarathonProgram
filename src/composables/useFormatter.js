import { format, parse } from 'date-fns';
import { fr } from 'date-fns/locale';
import { computed } from 'vue';
import { useI18n } from 'vue-i18n';

export const useFormatter = () => {
    const { locale } = useI18n();

    const formatter = computed(() => new Intl.NumberFormat(locale.value, { maximumSignificantDigits: 4 }));

    const capitalize = (string) => string.charAt(0).toUpperCase() + string.slice(1);

    const parseDate = (dateStr) => {
        return parse(dateStr, 'dd/MM/yyyy', new Date(), { locale: fr });
    };

    const getHourFromDate = (dateStr) => dateStr.slice(11, 16);

    const humanizeDate = (date) =>
        capitalize(format(date, 'eeee dd MMMM', locale.value === 'fr' ? { locale: fr } : {}));

    const formatNumber = (number, length) => {
        return String(number).padStart(length, '0');
    };

    const convertTimeToSeconds = (time) => {
        if (time.split(':').length === 2) {
            const [minutes, seconds] = time.split(':').map(Number);
            return minutes * 60 + seconds;
        } else if (time.split(':').length === 3) {
            const [hours, minutes, seconds] = time.split(':').map(Number);
            return hours * 3600 + minutes * 60 + seconds;
        }
    };

    const formatSecondsToMinutes = (seconds) => {
        const totalSeconds = Math.round(seconds);
        const minutes = Math.floor(totalSeconds / 60);
        const remainingSeconds = totalSeconds % 60;

        return `${formatNumber(minutes, 2)}'${formatNumber(remainingSeconds, 2)}"`;
    };

    return {
        formatter,
        humanizeDate,
        parseDate,
        formatNumber,
        convertTimeToSeconds,
        formatSecondsToMinutes,
        getHourFromDate,
    };
};
