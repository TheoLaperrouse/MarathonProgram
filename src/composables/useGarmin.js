import { ref } from 'vue';

export const useGarmin = () => {
    const isValidGarmin = ref(false);
    const formatTrainingToGarmin = (workout) => {
        return {
            name: workout.title,
            type: workout.type,
            duration: workout.duration,
            distance: workout.distance,
        };
    };
    const sendWorkoutToGarmin = () => {
        return;
    };
    return { formatTrainingToGarmin, sendWorkoutToGarmin, isValidGarmin };
};
