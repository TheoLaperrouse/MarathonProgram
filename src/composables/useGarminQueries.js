import { sendWorkout } from '@/services/garmin.service';
import { useMutation } from '@tanstack/vue-query';

export function useSendWorkout(accessToken) {
    return useMutation({
        mutationFn: (workoutData) => sendWorkout(accessToken, workoutData),
    });
}
