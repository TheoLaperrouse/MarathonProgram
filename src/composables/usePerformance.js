import { useFormatter } from '@/composables/useFormatter';
import { useLocalStorage } from '@vueuse/core';
import { computed } from 'vue';

export const usePerformance = () => {
    const { formatSecondsToMinutes, convertTimeToSeconds } = useFormatter();

    const cardiacFrequencyMax = 220;

    const bestTime = useLocalStorage('bestTime', '8:30');
    const marathonTime = useLocalStorage('marathonTime', '4:00:00');

    const bestTimeSeconds = computed(() => convertTimeToSeconds(bestTime.value));
    const marathonTimeSeconds = computed(() => convertTimeToSeconds(marathonTime.value));
    const VMAKmH = computed(() => (1 / (bestTimeSeconds.value / 2)) * 3600);
    const VO2Max = computed(() => parseInt(VMAKmH.value * 3.5));

    const getPaceByPercentVMA = (percent) => {
        const speed = VMAKmH.value * (percent / 100);
        const timePerKm = 60 / speed;
        return formatSecondsToMinutes(timePerKm * 60);
    };

    const getCardiacFrequencyByPercentVMA = (percent) => {
        return (cardiacFrequencyMax * percent) / 100;
    };

    const paces = computed(() => ({
        mediumRun: {
            pace: getPaceByPercentVMA(70),
            cardiacFrequency: getCardiacFrequencyByPercentVMA(70),
            percentVMA: 70,
            color: '#009848',
        },
        longRun: {
            pace: getPaceByPercentVMA(65),
            cardiacFrequency: getCardiacFrequencyByPercentVMA(60),
            percentVMA: 65,
            color: '#04C91A',
        },
        thresholdRun: {
            pace: getPaceByPercentVMA(82.5),
            cardiacFrequency: getCardiacFrequencyByPercentVMA(80),
            percentVMA: 82.5,
            color: '#77AABE',
        },
        intervalsRun: {
            pace: getPaceByPercentVMA(100),
            cardiacFrequency: getCardiacFrequencyByPercentVMA(80),
            percentVMA: 100,
            color: '#E12117',
        },
        fastIntervalsRun: {
            pace: getPaceByPercentVMA(110),
            cardiacFrequency: getCardiacFrequencyByPercentVMA(85),
            percentVMA: 110,
            color: '#E12117',
        },
        marathon: {
            pace: getPaceByPercentVMA(80),
            cardiacFrequency: getCardiacFrequencyByPercentVMA(80),
            percentVMA: 80,
            color: '#FFFF00',
        },
    }));

    return { paces, VO2Max, VMAKmH, marathonTime, bestTime, bestTimeSeconds, marathonTimeSeconds };
};
