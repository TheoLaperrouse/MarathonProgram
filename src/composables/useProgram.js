import { useFormatter } from '@/composables/useFormatter';
import {
    fourTrainingsDayProgram,
    fiveTrainingsDayProgram,
    sixTrainingsDayProgram,
} from '@/trainingPrograms/trainingPrograms';
import { useLocalStorage } from '@vueuse/core';
import { format, addDays, addMonths, subWeeks, getDay } from 'date-fns';
import { computed } from 'vue';

export const useProgram = () => {
    const matchProgram = {
        4: fourTrainingsDayProgram,
        5: fiveTrainingsDayProgram,
        6: sixTrainingsDayProgram,
    };

    const { humanizeDate } = useFormatter();

    const marathonDate = useLocalStorage('marathonDate', addMonths(new Date(), 3));
    const trainingDays = useLocalStorage('trainingDays', 4);
    const trainingDayChoices = useLocalStorage('trainingDayChoices', [0, 2, 4, 6]);

    const marathonProgramDate = computed(() => subWeeks(marathonDate.value, 12));

    const program = computed(() => matchProgram[trainingDays.value]);

    const formattedMarathonDate = computed(() => (marathonDate.value ? humanizeDate(marathonDate.value) : null));
    const formattedProgramDate = computed(() =>
        marathonProgramDate.value ? humanizeDate(marathonProgramDate.value) : null,
    );

    const getNextTrainingDay = (currentDate) => {
        const currentDay = getDay(currentDate);
        for (let i = 1; i <= 7; i++) {
            const nextDay = (currentDay + i) % 7;
            if (trainingDayChoices.value.includes(nextDay)) {
                return addDays(currentDate, i);
            }
        }
        return addDays(currentDate, 1);
    };

    const trainingSchedule = computed(() => {
        let currentDate = marathonProgramDate.value;
        const currentProgram = program.value;
        const schedule = {};

        Object.keys(currentProgram).forEach((weekKey) => {
            const sessions = currentProgram[weekKey];
            sessions.forEach((session) => {
                currentDate = getNextTrainingDay(currentDate);
                const dateStr = format(currentDate, 'dd/MM/yyyy');
                schedule[dateStr] = session;
            });
        });

        return Object.entries(schedule)
            .sort(
                ([dateA], [dateB]) =>
                    new Date(dateA.split('/').reverse().join('-')) - new Date(dateB.split('/').reverse().join('-')),
            )
            .reduce((sorted, [date, training]) => {
                sorted[date] = training;
                return sorted;
            }, {});
    });

    const dayTraining = computed(() => {
        const todayFormatted = format(Date.now(), 'dd/MM/yyyy');
        return trainingSchedule.value[todayFormatted];
    });

    return {
        dayTraining,
        marathonDate,
        formattedMarathonDate,
        trainingDays,
        trainingDayChoices,
        program,
        marathonProgramDate,
        trainingSchedule,
        formattedProgramDate,
    };
};
