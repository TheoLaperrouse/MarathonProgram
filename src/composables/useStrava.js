import { useGetAthleteData, useGetActivities } from '@/composables/useStravaQueries';
import { useLocalStorage } from '@vueuse/core';
import { parseISO, isToday } from 'date-fns';
import { computed } from 'vue';

export const useStrava = () => {
    const stravaToken = useLocalStorage('stravaToken', '');

    const { data: athleteData, isLoading: isAthleteLoading } = useGetAthleteData(stravaToken);
    const { data: activities, isLoading: isActivitiesLoading } = useGetActivities(stravaToken, { per_page: 10 });

    const isValidToken = computed(() => !!athleteData.value);

    const fullName = computed(() =>
        athleteData.value ? `${athleteData.value.firstname} ${athleteData.value.lastname}` : '',
    );

    const dayActivity = computed(() => {
        return activities.value?.filter((activity) => isToday(parseISO(activity.start_date_local)))[0];
    });

    return {
        athleteData,
        isAthleteLoading,
        activities,
        isActivitiesLoading,
        stravaToken,
        fullName,
        dayActivity,
        isValidToken,
    };
};
