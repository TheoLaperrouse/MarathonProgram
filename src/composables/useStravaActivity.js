import { useFormatter } from '@/composables/useFormatter';
import { computed } from 'vue';

export const useStravaActivity = (activity) => {
    const { getHourFromDate, formatSecondsToMinutes, humanizeDate } = useFormatter();
    const activityName = computed(() => activity.value.name);
    const activityTime = computed(() => formatSecondsToMinutes(activity.value.moving_time));
    const activityDistance = computed(() => activity.value.distance / 1000);
    const activityAveragePace = computed(() =>
        formatSecondsToMinutes(activity.value.moving_time / activityDistance.value),
    );
    const activityDate = computed(() => humanizeDate(activity.value.start_date_local));
    const activityStartTime = computed(() => getHourFromDate(activity.value.start_date_local));

    return { activityName, activityAveragePace, activityTime, activityDistance, activityDate, activityStartTime };
};
