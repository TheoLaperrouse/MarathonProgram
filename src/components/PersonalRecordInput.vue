<template>
    <div class="ml-3 mt-3 max-w-sm space-y-4 rounded-xl bg-gray-200 p-6 shadow-md">
        <h2 class="text-xl font-semibold text-gray-700">{{ $t('bestTimeInput') }}</h2>
        <input
            v-model="bestTime"
            type="number"
            placeholder="Enter your best 2km time"
            class="w-full rounded-md border border-gray-300 px-4 py-2 focus:outline-none focus:ring-2 focus:ring-indigo-500"
        />
    </div>
</template>

<script setup>
import { usePerformance } from '@/composables/usePerformance';

const { bestTime } = usePerformance();
</script>
