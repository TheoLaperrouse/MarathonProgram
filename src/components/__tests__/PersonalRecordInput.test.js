import PersonalRecordInput from '@/components/PersonalRecordInput.vue';
import { usePerformance } from '@/composables/usePerformance';
import { mount } from '@vue/test-utils';
import { ref } from 'vue';

vi.mock('vue-i18n', () => ({
    useI18n: () => ({
        t: (key) => key,
    }),
}));

vi.mock('@/composables/usePerformance');

describe('PersonalRecordInput', () => {
    let bestTimeMock;

    beforeEach(() => {
        vi.resetAllMocks();

        bestTimeMock = ref('7:15');
        usePerformance.mockReturnValue({ bestTime: bestTimeMock });
    });

    it('should match snapshot', () => {
        const wrapper = mount(PersonalRecordInput);

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders correctly with initial state', () => {
        const wrapper = mount(PersonalRecordInput);

        const input = wrapper.find('input');
        expect(input.element.value).toBe('7:15');
    });

    it('updates bestTime when input value changes', async () => {
        const wrapper = mount(PersonalRecordInput);

        const input = wrapper.find('input');
        await input.setValue('7:30');

        expect(bestTimeMock.value).toBe('7:30');
    });
});
