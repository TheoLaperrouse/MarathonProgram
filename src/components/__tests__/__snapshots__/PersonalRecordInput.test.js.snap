// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PersonalRecordInput > should match snapshot 1`] = `
"<div class="ml-3 mt-3 max-w-sm space-y-4 rounded-xl bg-gray-200 p-6 shadow-md">
  <h2 class="text-xl font-semibold text-gray-700">bestTimeInput</h2><input>
</div>"
`;
